code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/RegionMonitor.java

```diff
@@ -105,24 +134,56 @@ public void run() {
                             storedUUID!=null && !storedUUID.equals(externalProject.getUuid()));
                     if(uuidsDoNotMatch) {
                             objectProcessManager.executeStandardProcess(StandardProcess.REMOVE, link, null);
+                            continue;
                     }
                 }
+                if(link.getExternal()) {
+                        Account localAccount = objectManager.loadResource(Account.class, link.getAccountId());
+                        ExternalAccountLink accLink = RegionUtil.getAccountLinkForExternal(targetRegion, externalProject, localAccount, jsonMapper);
+                        if(accLink == null || removedStates.contains(accLink.getState())){
+                            objectProcessManager.executeStandardProcess(StandardProcess.REMOVE, link, null);
+                        }
+                } else {
+                    existingLinks.add(externalProjectKey);
+                }
             } catch (Exception ex) {
-                log.warn("Failed to monitor account link for %s - %s", link.getId(), ex);
+                log.warn(String.format("Failed to monitor account link for %s - %s", link.getId(), ex));
+                if(!link.getExternal()) {
+                        existingLinks.add(String.format("%s:%s", link.getLinkedRegion(), link.getLinkedAccount()));
+                }
+                return;
             }
         }
-        List<Account> accounts = objectManager.find(Account.class, ACCOUNT.KIND, AccountConstants.PROJECT_KIND, 
-            ACCOUNT.REMOVED, new Condition(ConditionType.NULL));
-        for(Account account : accounts) {
-            try {
-                regionService.reconcileExternalLinks(account.getId());
-            } catch (Exception ex) {
-                log.warn("Failed to reconcile external link for %s - %s", account.getId(), ex);
+    }
+    
+    private void cleanAgents(HashSet<String> existingLinks) {
+        List<Agent> agents = objectManager.find(Agent.class, AGENT.REMOVED, new Condition(ConditionType.NULL),
+                AGENT.EXTERNAL_ID, new Condition(ConditionType.NOTNULL));
+        
+        if(agents.size() > 0) { 
+            for(Agent agent : agents) {
+                String[] uri = agent.getUri().substring(EXTERNAL_AGENT_URI_PREFIX.length()).split("_");
+                String region_env = String.format("%s:%s", uri[0], uri[1]);
+                if(existingLinks.contains(region_env)) {
+                    continue; 
+                }
+                List<AccountLink> links = objectManager.find(AccountLink.class, ACCOUNT_LINK.LINKED_REGION, uri[0], 
+                        ACCOUNT_LINK.LINKED_ACCOUNT, uri[1], ACCOUNT_LINK.REMOVED, null);
+                if(links.size()>0) {
+                    continue;
+                }
+                try {
+                    objectProcessManager.executeStandardProcess(StandardProcess.DEACTIVATE, agent, null);
```

Comment:
it is possible to schedule a chain process; there should be a way for both deactivate and remove (deactivateThenRemove)